src/Data/HashMap/Linear.hs

```diff
@@ -216,62 +216,65 @@ insertAll ((k, v) : xs) hmap = insertAll xs (insert hmap k v)
 -- a full hashmap, return a probe result: the place the key already
 -- exists, a place to swap from, or an unfilled cell to write over.
 probeFrom :: Keyed k =>
-  (k, PSL) -> Int -> HashMap k v #-> (HashMap k v, ProbeResult k)
-probeFrom (k, p) ix (HashMap sz@(Size s) ct arr) = read arr ix & \case
-  (arr', Ur (PSL (-1), _)) ->
-    (HashMap sz ct arr', IndexToInsert (k, p) ix)
-  (arr', Ur (psl,(k',_))) -> case k == k' of
-    -- Note: in the True case, we must have p == psl
-    True -> (HashMap sz ct arr', IndexToUpdate (k, psl) ix)
-    False -> case psl < p of
-      True -> (HashMap sz ct arr', IndexToSwap (k, p) ix)
-      False -> probeFrom (k, p+1) ((ix+1)`mod` s) (HashMap sz ct arr')
-
+  (k, PSL) -> Int -> HashMap k v #-> (HashMap k v, ProbeResult k v)
+probeFrom (k, p) ix (HashMap ct arr) = Array.read arr ix & \case
+  (arr', Ur Nothing) ->
+    (HashMap ct arr', IndexToInsert p ix)
+  (arr', Ur (Just robinVal'@(RobinVal psl k' v'))) ->
+    case k == k' of
+      -- Note: in the True case, we must have p == psl
+      True -> (HashMap ct arr', IndexToUpdate v' psl ix)
+      False -> case psl < p of
+        True -> (HashMap ct arr', IndexToSwap robinVal' p ix)
```

Comment:
See this is confusing. The PSL `p` is of the query but in the `IndexToUpdate` constructor, the PSL is of the cell already there. Also, the `RobinVal` is of the current existing cell. All of this makes sense, it should just be documented in comments on the constructors.

Reply:
So, the `psl` on the constructurs are always the `psl` of the query. In `IndexToUpdate` case, they also happen to be the same as `psl` of the cell too.

But yeah, having different `psl`'s confusing. I'll document them.